src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs

```diff
@@ -129,22 +133,30 @@ public static int IndexOf(ref byte searchSpace, int searchSpaceLength, ref byte
                 {
                     Vector128<byte> cmpCh1 = Vector128.Equals(ch1, Vector128.LoadUnsafe(ref searchSpace, (nuint)offset));
                     Vector128<byte> cmpCh2 = Vector128.Equals(ch2, Vector128.LoadUnsafe(ref searchSpace, (nuint)(offset + ch1ch2Distance)));
-                    uint mask = (cmpCh1 & cmpCh2).AsByte().ExtractMostSignificantBits();
+                    Vector128<byte> cmpAnd = (cmpCh1 & cmpCh2).AsByte();
 
-                    while (mask != 0)
+                    // On Platforms with SSE41 and ARM64 use fast "is all zero" check
+                    // it's especially important for ARM64 where ExtractMostSignificantBits is expensive
+                    // but it also shows nice numbers on XArch
+                    bool useFastAllZeroCheck = Sse41.IsSupported || AdvSimd.Arm64.IsSupported;
```

Comment:
I wouldn't bother with the `useFastAllZeroCheck`.

For x86/x64, SSE4.1 is a 15 year old ISA and is old enough that even x86/x64 emulation on ARM64 (for both Windows and Mac) are able to provide it. The potential for a two instruction slower comparison on these paths for users running on extremely dated hardware isn't worth the additional complexity IMO.